lib/acl.c

```diff
@@ -0,0 +1,224 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "gatekeeper_acl.h"
+#include "gatekeeper_lls.h"
+
+/* Maximum number of rules installed per ACL. */
+#define MAX_NUM_IPV6_ACL_RULES (16)
+
+/*
+ * All IPv6 fields involved in classification; not all fields must
+ * be specified for every rule. Fields must be grouped into sets of
+ * four bytes, except for the first field.
+ */
+struct rte_acl_field_def ipv6_defs[NUM_FIELDS_IPV6] = {
+	{
+		.type = RTE_ACL_FIELD_TYPE_BITMASK,
+		.size = sizeof(uint8_t),
+		.field_index = PROTO_FIELD_IPV6,
+		.input_index = PROTO_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, proto),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST1_FIELD_IPV6,
+		.input_index = DST1_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[0]),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST2_FIELD_IPV6,
+		.input_index = DST2_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[4]),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST3_FIELD_IPV6,
+		.input_index = DST3_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[8]),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST4_FIELD_IPV6,
+		.input_index = DST4_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[12]),
+	},
+	{
+		/* Enforce grouping into four bytes. */
+		.type = RTE_ACL_FIELD_TYPE_BITMASK,
+		.size = sizeof(uint32_t),
+		.field_index = TYPE_FIELD_ICMPV6,
+		.input_index = TYPE_FIELD_ICMPV6,
+		.offset = sizeof(struct ipv6_hdr) +
+			offsetof(struct icmpv6_hdr, type),
+	},
+};
+
+int
+register_ipv6_acl(struct ipv6_acl_rule *ipv6_rules, unsigned int num_rules,
+	acl_cb_func cb_f, struct gatekeeper_if *iface)
+{
+	unsigned int numa_nodes = get_net_conf()->numa_nodes;
+	unsigned int i;
+
+	if (iface->acl_count == GATEKEEPER_IPV6_ACL_MAX) {
+		RTE_LOG(ERR, GATEKEEPER, "acl: cannot install more rules on the %s iface\n",
+			iface->name);
+		return -1;
+	}
+
+	/* Assign a new ID for this rule type. */
+	for (i = 0; i < num_rules; i++)
+		ipv6_rules[i].data.userdata = iface->acl_count;
+
+	for (i = 0; i < numa_nodes; i++) {
+		int ret = rte_acl_add_rules(iface->ipv6_acls[i],
+			(struct rte_acl_rule *)ipv6_rules, num_rules);
```

Comment:
I'm not sure why the ACL library does it this way, but `struct ipv6_acl_rule` is actually defined using a macro in include/gatekeeper_acl.h:

`RTE_ACL_RULE_DEF(ipv6_acl_rule, RTE_DIM(ipv6_defs));`

The `struct rte_acl_rule` is defined the same way (ctags won't find it either), but has no fields.